### PR TITLE
Feature/enhance uds aux with a start and stop tester present

### PR DESCRIPTION
**Pull request by TedRio**

### CODE REVIEW

File: src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py [hunk 1]
```diff
@@ -41,6 +44,38 @@ class UdsAuxiliary(UdsBaseAuxiliary):
...
+        uds_layer: dict = None,
+        **kwargs,
+    ):
+        """Initialize attributes.
+
+        :param com: communication channel connector.
+        :param config_ini_path: UDS parameter file.
+        :param odx_file_path: ecu diagnostic definition file.
+        :param request_id: optional CAN ID used for sending messages.
+        :param response_id: optional CAN ID used for receiving messages.
```
Comment (Pog3k) [inline src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py:64]: doc param tp_layer uds_layer missing

File: src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py [hunk 2]
```diff
@@ -271,6 +306,31 @@ def tester_present_sender(self, period: int = 4) -> Iterator[None]:
             stop_event.set()
             sender.join()
 
+    def start_tester_present_sender(self, period: int) -> None:
```
Comment (Pog3k) [inline src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py:309]: doesnt uds have a default period for tester present ?

File: examples/test_uds/test_raw_uds.py [hunk 3]
```diff
@@ -67,6 +67,12 @@ def test_run(self):
             # Go into safety system session
             uds_aux.send_uds_raw([0x10, 0x04])
 
+        # It is also possible to do it with start and stop methods
+        uds_aux.start_tester_present_sender(5)
```
Comment (Pog3k) [inline examples/test_uds/test_raw_uds.py:71]:
```suggestion
        uds_aux.start_tester_present_sender(period = 5)
```

File: src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py [hunk 4]
```diff
@@ -271,6 +306,31 @@ def tester_present_sender(self, period: int = 4) -> Iterator[None]:
...
+        )
+        self.sender.start()
+
+    def stop_tester_present_sender(self) -> None:
+        """Stop to continuously sends tester present messages via UDS"""
+        try:
+            self.sender_stop_event.set()
+            self.sender.join()
+            self.sender_stop_event.clear()
+        except AttributeError:
```
Comment (Pog3k) [inline src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py:329]: mhm.. why not if self.sender_stop_event.is_alive() ?

File: src/pykiso/lib/robot_framework/uds_auxiliary.py [hunk 5]
```diff
@@ -178,19 +175,13 @@ def start_tester_present_sender(self, period: int, aux_alias) -> None:
         :param aux_alias: auxiliary's alias
         """
         aux = self._get_aux(aux_alias)
-        self.sender = threading.Thread(
-            name="TesterPresentSender",
-            target=aux._sender_run,
-            args=(period, self.sender_stop_event),
-        )
-        self.sender.start()
+        aux.start_tester_present_sender(period)
```
Comment (Pog3k) [inline src/pykiso/lib/robot_framework/uds_auxiliary.py:178]: why is start_tester_present_sender not a lib function like  def stop_tester_present_sender ??

File: tests/test_robot_uds_auxiliary.py [hunk 6]
```diff
@@ -155,7 +155,7 @@ def test_tester_present_sender(mocker, robot_uds_aux, uds_aux):
     mocker.patch("time.sleep", return_value=None)
 
     robot_uds_aux.start_tester_present_sender(1, "uds_aux")
```
Comment (Pog3k) [inline tests/test_robot_uds_auxiliary.py:157]:
```suggestion
    robot_uds_aux.start_tester_present_sender(period = 1, "uds_aux")
```

File: docs/examples/uds_auxiliary.rst [hunk 7]
```diff
@@ -283,3 +285,15 @@ UDS tester present sender
...
+
+|It is also possible to start and stop the tester present sender manually with the methods
+|start_tester_present_sender and stop_tester_present_sender.
+|Warning: if start_tester_present_sender is called but not stop_tester_present_sender the
+|tester present sender thread will still be running
+
+.. code:: python
+
+    # start sending tester present messages every 3 seconds until the context manager is exited
+    uds_aux.start_tester_present_sender(1)
```
Comment (Pog3k) [inline docs/examples/uds_auxiliary.rst:297]:
```suggestion
    uds_aux.start_tester_present_sender(period=1)
```

**Review by sebastianpfischer — CHANGES_REQUESTED** [review]
Sorry, I do not understand the PR...
It is a duplicate of what we already added by https://github.com/eclipse/kiso-testing/pull/87/files

I would recommend you to read https://docs.python.org/3/library/contextlib.html#contextlib.AbstractContextManager

Based on that, Python already have 2 special methods implemented in the background that does what you want to do:
`tester_present_sender.__enter__()`
`tester_present_sender.__exit__()`

No additional threads or events are here needed.
The tricky part is indeed when the test failed and the stop is not called.
To tackle this, https://github.com/eclipse/kiso-testing/blob/master/src/pykiso/lib/auxiliaries/udsaux/common/uds_base_auxiliary.py#L149 should probably be overwritten to ensure a graceful shutdown.

File: src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py (hunk 4 above)
Comment (Pog3k) [reply src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py:329]: ah ok,  But what happens when the user calls 
start_tester_present_sender    multiple times ?
Does the unstopped thread then run in the background ?

File: src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py (hunk 1 above)
Comment (TedRio) [reply src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py]: the init has now been deleted

File: src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py (hunk 2 above)
Comment (TedRio) [reply src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py]: done

File: examples/test_uds/test_raw_uds.py (hunk 3 above)
Comment (TedRio) [reply examples/test_uds/test_raw_uds.py]: Done

File: tests/test_robot_uds_auxiliary.py (hunk 6 above)
Comment (TedRio) [reply tests/test_robot_uds_auxiliary.py]: Done

File: docs/examples/uds_auxiliary.rst (hunk 7 above)
Comment (TedRio) [reply docs/examples/uds_auxiliary.rst]: Changed

**Comment by TedRio** [comment]
> Sorry, I do not understand the PR... It is a duplicate of what we already added by https://github.com/eclipse/kiso-testing/pull/87/files
> 
> I would recommend you to read https://docs.python.org/3/library/contextlib.html#contextlib.AbstractContextManager
> 
> Based on that, Python already have 2 special methods implemented in the background that does what you want to do: `tester_present_sender.__enter__()` `tester_present_sender.__exit__()`
> 
> No additional threads or events are here needed. The tricky part is indeed when the test failed and the stop is not called. To tackle this, https://github.com/eclipse/kiso-testing/blob/master/src/pykiso/lib/auxiliaries/udsaux/common/uds_base_auxiliary.py#L149 should probably be overwritten to ensure a graceful shutdown.


-> at the end I did manage to keep this context manager and have the expected behaviour without using __enter__() and exit()

File: src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py (hunk 4 above)
Comment (TedRio) [reply src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py:289]: Should not happened now

**Review by sebastianpfischer — CHANGES_REQUESTED** [review]
Almost there!

**Comment by sebastianpfischer** [comment]
Promise, it will be my last comments :)

**Review by sebastianpfischer — APPROVED** [review]

File: src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py [hunk 8]
```diff
@@ -293,3 +344,12 @@ def _receive_message(self, timeout_in_s: float) -> None:
     def _run_command(self, cmd_message, cmd_data=None) -> Union[dict, bytes, bool]:
         """Not used."""
         pass
+
+    def _delete_auxiliary_instance(self) -> bool:
+        """Close current associated channel.
```
Comment (Pog3k) [inline src/pykiso/lib/auxiliaries/udsaux/uds_auxiliary.py:349]: Which channel ? (:
.>stop tester present if active

**Review by Pog3k — APPROVED** [review]

**Pull request merged**